Rendering of inner constructs with domain directives and improved Sphinx 1.3 support

vitaut opened the pull request:
The proposed PR implements rendering of inner language constructs, such as classes, with domain directives. To achieve this, the domain directive factory (which is now an object, not a dictionary) is propagated to renderers and used to create domain directives whenever we need to render something. Constructs such as files that are not part of the language are rendered with a new class, `FileRenderer`.

Enums, typedefs, unions and namespaces are rendered using `:cpp:type` directive as Sphinx doesn't have designated directives. This is probably OK for the first three as they are actually types, but not great for namespaces. This might be something we'd like to contribute to Sphinx in the future.

Rendering of functions is temporary back to the old state (without using Sphinx directives) as adding function support would make the PR even bigger. I plan to add this in the next PR which will complete integration with Sphinx and make all Breathe features work with both Sphinx 1.2 and 1.3.

michaeljones commented:
Wow :) An impressive list! I'll do my best to review it in the morning. Sounds like a very positive step though!

michaeljones commented:
Hey, sorry, just a couple of minor comments there but nothing serious. Thanks very much for such a clean line of development with separate commits and an understandable progression. It is really exciting to see.

The 'compare' diff looks all positive. 

If I can find the time I might make a few opinionated changes but I'd do it in a separate PR and ask you to review them.

Merge away! And thanks again. Great to see.

vitaut commented:
I've addressed your comments and merged this. I'm pretty sure the code can be improved in a number of ways, so please do whatever changes you see beneficial. Thanks for reviewing!
